stl/inc/xutility

```diff
@@ -3802,63 +3888,111 @@ public:
-    -> decltype(_Left.base() - _Right.base()) {
-    return _Left.base() - _Right.base();
-}
-
 template <class _Iter1, class _Iter2>
-_NODISCARD _CONSTEXPR17 bool operator==(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right) {
-    return _Left.base() == _Right.base();
-}
+_NODISCARD _CONSTEXPR17 bool operator==(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right)
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    requires requires {
+        { _Left.base() == _Right.base() } -> _Implicitly_convertible_to<bool>;
+    }
+// clang-format on
+#endif // __cpp_lib_concepts
+{ return _Left.base() == _Right.base(); }
 
+#if !_HAS_CXX20 || __cpp_impl_three_way_comparison < 201902L
 template <class _Iter1, class _Iter2>
 _NODISCARD _CONSTEXPR17 bool operator!=(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right) {
     return !(_Left == _Right);
 }
+#endif // !_HAS_CXX20 || __cpp_impl_three_way_comparison < 201902L
 
 template <class _Iter1, class _Iter2>
-_NODISCARD _CONSTEXPR17 bool operator<(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right) {
-    return _Left.base() < _Right.base();
-}
+_NODISCARD _CONSTEXPR17 bool operator<(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right)
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    requires requires {
+        { _Left.base() < _Right.base() } -> _Implicitly_convertible_to<bool>;
+    }
+// clang-format on
+#endif // __cpp_lib_concepts
+{ return _Left.base() < _Right.base(); }
 
 template <class _Iter1, class _Iter2>
-_NODISCARD _CONSTEXPR17 bool operator>(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right) {
-    return _Right < _Left;
-}
+_NODISCARD _CONSTEXPR17 bool operator>(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right)
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    requires requires { _Right < _Left; }
+// clang-format on
+#endif // __cpp_lib_concepts
+{ return _Right < _Left; }
+
+template <class _Iter1, class _Iter2>
+_NODISCARD _CONSTEXPR17 bool operator<=(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right)
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    requires requires { _Right < _Left; }
+// clang-format on
+#endif // __cpp_lib_concepts
+{ return !(_Right < _Left); }
 
 template <class _Iter1, class _Iter2>
-_NODISCARD _CONSTEXPR17 bool operator<=(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right) {
-    return !(_Right < _Left);
+_NODISCARD _CONSTEXPR17 bool operator>=(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right)
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    requires requires { _Left < _Right; }
+// clang-format on
+#endif // __cpp_lib_concepts
+{ return !(_Left < _Right); }
+
+#if defined(__cpp_impl_three_way_comparison) && defined(__cpp_lib_concepts)

```

Comment:
Thinking out loud. Should this have an else clause that encloses everything above from `operator==` onwards.

The individual definitions should not be necessary with spaceship

Reply:
`move_iterator`'s `<=>` is only provided when the adapted iterator provides `<=>`, i.e., it requires `three_way_comparable_with`. We still need these individual operators when the adaptee doesn't provide `<=>`, or when the adaptee has relational operators that are inconsistent with `operator<=>`.